Telephony (#1520)
oktadeveloper/okta-blog: _source/_posts/2024-08-15-otp-over-sms.md

---
layout: blog_post
title: "Approaches to keep sending OTP over SMS... for now"
author: maurice-sharp
by: contractor
communities: [devops,.net]
description: "Learn how to make the best telephony choices for your application"
tags: [devops]
tweets:
- ""
image: blog/telephony/social.jpg
type: conversion
---

> ## Approaches to keep sending OTP over SMS... for now

Reviewer: ```suggestion
{% include toc.md %}

## Approaches to keep sending OTP over SMS... for now
```

Sending messages to more than one region may have at least two impacts. First, check that your desired vendor or vendors cover those regions.

Second, the features and regulations for traffic may differ from region to region. Some of the differences include:

* Limitations on the types of entities that can send messages by SMS. This typically requires proof of identity and business registration.

* Registration of a sender ID for your business. For example, messages without a valid sender ID are automatically marked as "Likely-SCAM" in Singapore.

* Using *short codes*–special telephone numbers designed for high traffic. This can add significant cost.

* Supported formats, such as ASCII and Unicode.

* Character length limits for messages. Note that each Unicode item counts as two characters.

Check that your vendor supports the regulations in your desired regions.

### How many messages?

Telephony vendors or service providers need to know the volume of messages. And not just the average volume, but any peaks, such as a time when a majority of people are trying to sign on to your network.

> The service cost is the most obvious thing related to volume, but there are two others. First is the impact on the rate limits used to prevent spam texts. These limits can prevent your messages from being sent, especially during peak volume. Vendors may be able to increase limits.

Reviewer: ```suggestion
The service cost is the most obvious related to volume issues, but there are two others. First is the impact on the rate limits used to prevent spam texts. These limits can prevent messages from being sent, especially during peak volume; vendors may be able to increase limits.
```

---
layout: blog_post
title: "Approaches to keep sending OTP over SMS... for now"
author: maurice-sharp
by: contractor
communities: [devops,.net]
description: "Learn how to make the best telephony choices for your application"
tags: [devops]
tweets:
- ""
image: blog/telephony/social.jpg
type: conversion
---

> {% include toc.md %}

Reviewer 2: ```suggestion
```

Please move this below intro.

> ## Approaches to keep sending OTP over SMS... for now

Reviewer 2: ```suggestion
```

Please delete this header. This is the same as the title. A blog should not start immediately with a header as the title is a header. There needs to be an intro.

> "SMS has long played an important role as a universally applicable method of verifying a user's identity via one-time passcodes. And over the last decade, SMS and voice-based Multifactor Authentication has prevented untold attempts to compromise user accounts.

Reviewer 2: Consider using `>` plus citation tags. We have examples of it in the blog

But it's time to move on."

– Ben King, VP Customer Trust: [BYO Telephony and the future of SMS at Okta](https://sec.okta.com/articles/2023/08/byo-telephony-and-future-sms-okta?_gl=1*50v0v8*_gcl_au*NzM2MTA4NjA5LjE3MTk5NTY2MDU.*_ga*MTE1NjAwNzQxNC4xNjY0OTc5MTk3*_ga_QKMSDV5369*MTcyMjg4MjA0Ni4yODIuMS4xNzIyODgzNDM3LjYwLjAuMA..&_ga=2.145329393.498111759.1722873073-1156007414.1664979197)

## SMS/Voice is too SIMple

The one-time passcode (OTP) you send using SMS or Voice may not go to the phone you want. SIM swapping–stealing someone else's phone number–lets bad actors receive the message or call with the code. They're one step closer to breaking into your system. And if all it takes is an account name and OTP, they may succeed. And it's not just SIM hacking; other issues include:

* No phishing resistance

* No control of the channel for sending secrets

* No way to link a user to their device

* Longer login times than other methods

> Okta [recommended moving away](https://www.okta.com/blog/2020/05/why-you-should-ditch-sms-as-an-auth-factor/)[ from SMS/Voice authentication](https://www.okta.com/blog/2020/05/why-you-should-ditch-sms-as-an-auth-factor/) some time ago. There are many other factors you can use for authentication, including:

Reviewer 2: ```suggestion
Okta [recommended moving away](/blog/2020/05/why-you-should-ditch-sms-as-an-auth-factor/)[ from SMS/Voice authentication](/blog/2020/05/why-you-should-ditch-sms-as-an-auth-factor/) some time ago. There are many other factors you can use for authentication, including:
```

Blog references do not include domain so that preview mode works. Also, consider using the new embed tag

Reviewer 2: Will people be able to distinguish between the two links?

* Generating codes in an authenticator app such as Okta Verify, Authy, Google Authenticator, or 1Password.

> * FIDO2.0 (WebAuthn) which, in addition to phones, can use hardware keys and on-device authenticators.

Reviewer 2: Are you sure this is the correct term? (stating "FIDO2.0 (WebAuthn)") I'm not sure the parenthesis here is correct, probably should verify with Okta standards for reference?

Author: good question; Maurice went through other reviews on the content so I'll ask

> Soon, [Okta will](https://support.okta.com/help/s/article/bring-your-own-telephony-required-for-sms-and-voice?language=en_US)[ require you to bring your own telephony provider](https://support.okta.com/help/s/article/bring-your-own-telephony-required-for-sms-and-voice?language=en_US) to keep sending those codes. If you need time to move to a different method of verifying identity, you must configure your own provider for SMS/Voice.

Reviewer 2: ```suggestion
Soon, [Okta will](https://support.okta.com/help/s/article/bring-your-own-telephony-required-for-sms-and-voice)[ require you to bring your own telephony provider](https://support.okta.com/help/s/article/bring-your-own-telephony-required-for-sms-and-voice) to keep sending those codes. If you need time to move to a different method of verifying identity, you must configure your own provider for SMS/Voice.
```

To allow for reader's language settings

## Hooked on telephony

> You can send the OTP in the SMS/Voice flow using the [telephony inline hook]([https://help.okta.com/oie/en-us/content/topics/telephony/telephony-inline-hook.htm](https://help.okta.com/oie/en-us/content/topics/telephony/telephony-inline-hook.htm)). Okta uses the code or URL in the hook to send the OTP, though, as you'll see, the hook may not be called every time (and that's a good thing). When your hook fails to send the message or takes too long to update the status, Okta takes over sending the message. However, the number of those messages is heavily rate-limited.

Reviewer 2: does a link to the page suffice? (meaning does it also need the entire URL written as well?)

> The code or URL you provide may simply send the message and communicate the outcome to Okta. The code or server may be more complex, managing geo-specific vendors, failure, failover to another provider, and hacking. No matter how easy or complex the code, there are three main approaches:

Reviewer 2: ```suggestion
The code or URL you provide may simply send the message and communicate the outcome to Okta. The code or server may be more complex, managing geo-specific vendors, failure, failover to another provider, and hacking. No matter the code, there are three main approaches:
```

I recommend steering away from these terms for inclusivity.

Author: good catch

1. Implement the code and use your own telephony provider or providers.

2. Outsource the implementation and use your own telephony provider or providers.

3. Use a managed service that manages the process for you.

Some of the main things to consider when choosing an approach are the regions for messages, the expected traffic, the desired reliability, branding requirements, protection from hacking, and your resources.

> ### Which regions?

Reviewer 2: Consider working keywords into the H2 and H3 headers for SEO. If that is at all a concern.

Two things can identify a region. First are any regulations for sending messages. Those regulations can be set by collectives, such as the European Union, countries, or even sub-parts of a country. Second is the area covered by the telco sending the message. 

Sending messages to more than one region may have at least two impacts. First, check that your desired vendor or vendors cover those regions.

Second, the features and regulations for traffic may differ from region to region. Some of the differences include:

* Limitations on the types of entities that can send messages by SMS. This typically requires proof of identity and business registration.

* Registration of a sender ID for your business. For example, messages without a valid sender ID are automatically marked as "Likely-SCAM" in Singapore.

> * Using *short codes*–special telephone numbers designed for high traffic. This can add significant cost.

Reviewer 2: ```suggestion
* Using *short codes*—special telephone numbers designed for high traffic. This can add significant cost.
```

Writing style is em dash

There are also accounts with the provider or service that must be secured. 

### How many people?

No matter the other concerns you identified, processes will change and update, and new things will need to be done.

New message flows and failovers require updating existing support processes for SMS/Voice users. This may include working with your chosen telephony or service vendor. You may also need to add more frequent log monitoring to detect when the failover rate limit prevents Okta from sending messages.

Vendors need management. Projects for implementing the chosen approach need planning and project management. The resources for the implementation phase vary significantly. 

Implementing custom code is similar to adding a somewhat complex feature to your product: it requires product management/specification, design, engineering, testing, and project management. Outsourcing the implementation can reduce the technical resources but adds vendor management.

Moving to a service provider minimizes the technical requirements, though there's still vendor management and monitoring.

> ## Designing a DIY Hook

Reviewer 2: ```suggestion
## Designing a DIY hook
```

## Vendors

The kind of vendors you need depends on your approach. Below are a few possibilities. Some are [recommendations from Okta](https://www.okta.com/okta-telecommunication-services/), and others are suggestions. No matter what, make sure that the vendor meets your criteria.

### Telephony providers

Here are some vendors you can use to implement the hook in-house or with a consultant.

* [Telesign](https://www.telesign.com/okta-telesign)

* [Twillio](https://pages.twilio.com/Twilio-Messaging-Okta.html)

* [Vonage](https://www.vonage.com/communications-apis/campaigns/okta/)

### Consultants

Many consulting companies can implement the hook for you. Another option is to use [Okta professional services](https://www.okta.com/services/professional-services/).

### Services

Some services deliver the SMS for you. That can include handling unavailable telephony vendors, resends, and other issues. Adding a service usually requires only adding a URL for the telephony hook.

Services include:

* [AWS Pinpoint](https://aws.amazon.com/pinpoint/)

* [BeyondID](https://beyondid.com/solutions/sms-for-okta/)

* [Twilio Verify](https://www.twilio.com/docs/verify/api)

> ## What Next? 

Reviewer 2: ```suggestion
## What's next? 
```

I still recommend working in keywords though... But headers are sentence case per the writing guidelines.
